WORKFLOW.md

```diff
@@ -0,0 +1,50 @@
+_below is more like open disscussion than a design doc_
+### Introduction
+
+The following CRDs are defined for managing the lifecycle of Buckets:
+
+ - BucketRequest - Represents a request to provision a Bucket
+ - BucketClass - Represents a class of Buckets with similar characteristics
+ - Bucket - Represents a Bucket or its equivalent in the storage backend
+
+ The following CRDs are defined for managing the lifecycle of workloads accessing the Bucket:
+
+ - BucketAccessRequest - Represents a request to access a Bucket
+ - BucketAccessClass - Represents a class of accessors with similar access requirements
+ - BucketAccess - Represents a access token or service account in the storage backend
+
+Simialr to CSI, it has three running components :
+- Controller - mananges life cycle of COSI objects, watches for the requests for above mentioned objects
+- Node adapter - bootstraps bucket access request to applications for their worloads
+- Sidecar/Driver - receives the request and calls corresponding api in the driver
+
+More details and defintions can be found the [COSI Spec](https://github.com/kubernetes-sigs/container-object-storage-interface-spec/blob/master/spec.md)
+
+### Workflow
+
+- BucketRequest(namespace scoped) can be created with reference to a BucketClass, so user will get a Bucket(clusterscoped). The COSI driver will create a `backend bucket` in RGW for a Bucket Request
+- BucketAccessRequest(namespace scoped) can be create with reference to a BucketAccessClass and BucketRequest, so user will get a BucketAccess(clusterscoped).The COSI driver will create a `s3 user` with proper access
+- 
+### COSI Driver
+The `user credentials` and `endpoint` need to passed as `Kubernetes Secret` to the Driver. Using that details, the driver will perform bucket operations via S3 apis and user operations via rados-admin Ops apis. 
+
+### Consuming at application
+The app pod can refer BucketAccessRequest and Node adapter details as volume mount as follows:
+```yaml
+spec:
+  containers:
+      volumeMounts:
+        - name: cosi-secrets
+          mountPath: /data/cosi
+  volumes:
+  - name: cosi-secrets
+    csi:
+      driver: objectstorage.k8s.io
+      volumeAttributes:
+        bar-name: sample-bar
+        bar-namespace: default
+```
+The adapter bootstrap the credentials into application pod, for Ceph COSI Driver credentials passed as `.aws/credentials` file into the pod.
+
+### Integrating with Rook
+The controller, node adapter and sidecar/driver need to deployed via Rook. Rook also need to create a `cephobjectstoreuser` and refer the secret in the deployment.
```

Comment:
> The controller, node adapter and sidecar/driver need to deployed via Rook.

Let's expand this section more. Here are some questions:
- What image(s) need to be deployed for the controller, node adapter, and sidecar/driver? 
   - Does K8s create images for COSI, like it does for CSI? Or is COSI a library and we need to build our own images? If we need to build the images what needs to be in the base image? I wonder if we just need a simple s3 client, and if we need the ceph base image.
- How many of each of those pods are needed? 
   - Is there just one controller, or are there two instances for HA, or perhaps configurable?
   - Is the node adapter a daemonset (on all nodes)?
   - Is the sidecar/driver deployed with the controller pod, node adapter pod, or independently?

> Rook also need to create a `cephobjectstoreuser` and refer the secret in the deployment.

Is this user for administration of all the buckets created by the object store? Or is there one user per bucket? Is there a naming convention for the secret? Is there something we can re-use from the OBC implementation?

Reply:
Thanks for the questions. I will try to answer the following according to my best knowledge
> > The controller, node adapter and sidecar/driver need to deployed via Rook.
> 
> Let's expand this section more. Here are some questions:
> 
>     * What image(s) need to be deployed for the controller, node adapter, and sidecar/driver?
>       
>       * Does K8s create images for COSI, like it does for CSI? Or is COSI a library and we need to build our own images? If we need to build the images what needs to be in the base image? I wonder if we just need a simple s3 client, and if we need the ceph base image.

As of now, there are different images for the controller, node adapter and sidecar/driver. I hope like CSI, the COSI will be part of k8S and we don't need to build those images our own and the driver image will come from ceph-cosi-repo. @jeffvance / @krishchow .
> 
>     * How many of each of those pods are needed?
>       
>       * Is there just one controller, or are there two instances for HA, or perhaps configurable?
>       * Is the node adapter a daemonset (on all nodes)?
>       * Is the sidecar/driver deployed with the controller pod, node adapter pod, or independently?

This part even I am not quite(abt no of pods) maybe we refer current CSI design and check with them for their ideas.  In my dev set up I used only one instance of each of them didn't experiment much with it
AFAIK the node adapter needs to be present on all the nodes. The sidecar/driver deployed independently

> 
> 
> > Rook also need to create a `cephobjectstoreuser` and refer the secret in the deployment.
> 
> Is this user for administration of all the buckets created by the object store? Or is there one user per bucket? Is there a naming convention for the secret? Is there something we can re-use from the OBC implementation?

In the current code, the User is the kind of administrator who creates all the buckets as part of the `BucketRequest` it receives and also creates users on receiving `BucketAccessRequest`. But I tend to change this (more alike OBC implementation) create a new user for `BucketRequest` and create the bucket using that new user.